src/network/core/http_winhttp.cpp

```diff
@@ -0,0 +1,301 @@
+ *
+ * The callback needs to call some WinHttp functions for certain states, so WinHttp continues
+ * to read the request. This also allows us to abort when things go wrong, by simply not calling
+ * those functions.
+ * Comments with "Next step:" mark where WinHttp needs a call to continue.
+ *
+ * @param code    The code of the event.
+ * @param info    The information about the event.
+ * @param length  The length of the information.
+ */
+void NetworkHTTPRequest::WinHttpCallback(DWORD code, void *info, DWORD length)
+{
+	switch (code) {
+		case WINHTTP_CALLBACK_STATUS_RESOLVING_NAME:
+		case WINHTTP_CALLBACK_STATUS_NAME_RESOLVED:
+		case WINHTTP_CALLBACK_STATUS_CONNECTING_TO_SERVER:
+		case WINHTTP_CALLBACK_STATUS_CONNECTED_TO_SERVER:
+		case WINHTTP_CALLBACK_STATUS_SENDING_REQUEST:
+		case WINHTTP_CALLBACK_STATUS_REQUEST_SENT:
+		case WINHTTP_CALLBACK_STATUS_RECEIVING_RESPONSE:
+		case WINHTTP_CALLBACK_STATUS_RESPONSE_RECEIVED:
+		case WINHTTP_CALLBACK_STATUS_CLOSING_CONNECTION:
+		case WINHTTP_CALLBACK_STATUS_CONNECTION_CLOSED:
+		case WINHTTP_CALLBACK_STATUS_HANDLE_CREATED:
+		case WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING:
+			/* We don't care about these events, and explicitly ignore them. */
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_REDIRECT:
+			/* Make sure we are not in a redirect loop. */
+			if (this->depth++ > 5) {
+				Debug(net, 0, "HTTP request failed: too many redirects");
+				this->callback->OnFailure();
+				this->finished = true;
+				return;
+			}
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_SENDREQUEST_COMPLETE:
+			/* Next step: read response. */
+			WinHttpReceiveResponse(this->request, nullptr);
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_HEADERS_AVAILABLE:
+		{
+			/* Retrieve the status code. */
+			DWORD status_code = 0;
+			DWORD status_code_size = sizeof(status_code);
+			WinHttpQueryHeaders(this->request, WINHTTP_QUERY_STATUS_CODE | WINHTTP_QUERY_FLAG_NUMBER, WINHTTP_HEADER_NAME_BY_INDEX, &status_code, &status_code_size, WINHTTP_NO_HEADER_INDEX);
+			Debug(net, 3, "HTTP request status code: {}", status_code);
+
+			/* If there is any error, we simply abort the request. */
+			if (status_code >= 400) {
+				Debug(net, 0, "HTTP request failed: status-code {}", status_code);
+				this->callback->OnFailure();
+				this->finished = true;
+				return;
+			}
+
+			/* Next step: query for any data. */
+			WinHttpQueryDataAvailable(this->request, nullptr);
+		} break;
+
+		case WINHTTP_CALLBACK_STATUS_DATA_AVAILABLE:
+		{
+			/* Retrieve the amount of data available to process. */
+			DWORD size = *(DWORD *)info;
+
+			/* Next step: read the data in a temporary buffer. */
+			char *buffer = size == 0 ? nullptr : MallocT<char>(size);
+			WinHttpReadData(this->request, buffer, size, 0);
+		} break;
+
+		case WINHTTP_CALLBACK_STATUS_READ_COMPLETE:
+			Debug(net, 4, "HTTP callback: {} bytes", length);
+
+			this->callback->OnReceiveData(static_cast<char *>(info), length);
+			free(info);
```

Comment:
I'd at cross-reference that this is freeing the allocation that is made in the `WINHTTP_CALLBACK_STATUS_DATA_AVAILABLE` case in a comment.